torch/nativert/detail/Pytree.cpp

```diff
@@ -0,0 +1,485 @@
+    treeMapArgs(
+        TreeMapFn f,
+        const std::vector<c10::IValue>& args,
+        const std::unordered_map<std::string, c10::IValue>& kwargs,
+        const TreeSpec& spec) {
+  const auto val = argsToIValue(args, kwargs);
+  const auto mapVal = treeMap(f, val, spec);
+  auto mapArgs =
+      mapVal.toTupleRef().elements()[0].toTupleRef().elements().vec();
+  std::unordered_map<std::string, c10::IValue> mapKwargs;
+  for (const auto& entry : mapVal.toTupleRef().elements()[1].toGenericDict()) {
+    mapKwargs.emplace(entry.key().toStringRef(), entry.value());
+  }
+  return {std::move(mapArgs), std::move(mapKwargs)};
+}
+
+void leafApply(
+    TreeMapNoReturnFn fn,
+    const c10::IValue& tree,
+    const TreeSpec& spec) {
+  if (spec.isLeaf()) {
+    if (spec.isUsed()) {
+      fn(tree, spec.value());
+    }
+    return;
+  }
+  auto leafApplyFn = spec.nodeDefCache().leafApplyFn;
+  leafApplyFn(fn, tree, spec);
+}
+
+nlohmann::json defaultContextLoadFn(std::string_view context) {
+  return nlohmann::json::parse(context);
+}
+
+TreeSpec::TreeSpec(
+    std::string_view uniformName,
+    nlohmann::json context,
+    std::vector<TreeSpec> children,
+    NodeDef nodeDefCache)
+    : uniformName_(uniformName),
+      context_(std::move(context)),
+      children_(std::move(children)),
+      nodeDefCache_(nodeDefCache),
+      numLeaves_(0),
+      value_(nullptr),
+      isUsed_(false) {
+  for (auto& child : children_) {
+    numLeaves_ += child.numLeaves();
+    allLeaves_ &= child.isLeaf();
+    isUsed_ |= child.isUsed();
+  }
+
+  if (uniformName_ == "builtins.dict" ||
+      uniformName_ == "torch.fx.immutable_collections.immutable_dict") {
+    for (const auto& keyObj : context_) {
+      contextKeys_.push_back(dynamicToIValue(keyObj));
+    }
+  }
+}
+
+c10::TypePtr TreeSpec::toAtenType() const {
+  if (isLeaf()) {
+    return c10::AnyType::get();
+  } else if (uniformName_ == "builtins.tuple") {
+    std::vector<c10::TypePtr> childrenType;
+    childrenType.reserve(children_.size());
+    for (const auto& childrenSpec : children_) {
+      childrenType.emplace_back(childrenSpec.toAtenType());
+    }
+    return c10::TupleType::create(std::move(childrenType));
+  } else if (
+      uniformName_ == "builtins.list" ||
+      uniformName_ == "torch.fx.immutable_collections.immutable_list") {
+    if (children_.empty()) {
+      return c10::ListType::create(c10::AnyType::get());
+    } else {
+      return c10::ListType::create(children_[0].toAtenType());
+    }
+  } else if (
+      uniformName_ == "builtins.dict" ||
+      uniformName_ == "torch.fx.immutable_collections.immutable_dict") {
+    if (children_.empty()) {
+      return c10::DictType::create(c10::AnyType::get(), c10::AnyType::get());
+    } else {
+      return c10::DictType::create(
+          dynamicToIValue(context_[0]).type(), children_[0].toAtenType());
+    }
+  } else {
+    TORCH_CHECK(false, "Unsupported uniform name: ", uniformName());
+  }
+}
```

Comment:
We should support `PyStructSequence` types in `torch.return_types.*`.

Reply:
> We should support `PyStructSequence` types in `torch.return_types.*`.

We can add this as a followup. This is mainly used for the serialized input and output specs of exported program. Currently, there's no default serialization implemented for `torch.return_types.*`. The users would need to update their _register_pytree_node call with a `serialized_type_name` kwarg for these types. If we want to hard-code the  `uniformName` here for `torch.return_types.*`, it might be good to do that along with a default serialization implementation for exported program. 

cc @zhxchen17

Reply:
> If we want to hard-code the `uniformName` here for `torch.return_types.*`

`torch.return_types.*` are first-party types that we can hardcode the behaviors.

Reply:
> > If we want to hard-code the `uniformName` here for `torch.return_types.*`
> 
> `torch.return_types.*` are first-party types that we can hardcode the behaviors.

@XuehaiPan yeah, I'm just saying if we hard code it here, we need to hard code it in the exported program serializer as well. And I think it might be good to add and test this in a separate PR. In this PR we're just moving the code to OSS. We can extend the code as a follow up.